src/xrpld/rpc/handlers/Simulate.cpp

```diff
@@ -0,0 +1,307 @@
+//   binary: <bool>
+// }
+Json::Value
+doSimulate(RPC::JsonContext& context)
+{
+    context.loadType = Resource::feeMediumBurdenRPC;
+
+    Json::Value jvResult;  // the returned result
+    Json::Value tx_json;   // the tx as a JSON
+
+    // check validity of `binary` param
+    if (context.params.isMember(jss::binary))
+    {
+        auto const binary = context.params[jss::binary];
+        if (!binary.isBool())
+        {
+            return RPC::invalid_field_error(jss::binary);
+        }
+    }
+
+    if (context.params.isMember(jss::tx_blob))
+    {
+        if (context.params.isMember(jss::tx_json))
+        {
+            // both `tx_blob` and `tx_json` included
+            return rpcError(rpcINVALID_PARAMS);
+        }
+        auto const blob = context.params[jss::tx_blob];
+        if (!blob.isString())
+        {
+            return RPC::invalid_field_error(jss::tx_blob);
+        }
+        auto unHexed = strUnHex(blob.asString());
+
+        if (!unHexed || !unHexed->size())
+            return RPC::invalid_field_error(jss::tx_blob);
+        try
+        {
+            SerialIter sitTrans(makeSlice(*unHexed));
+            tx_json = STObject(std::ref(sitTrans), sfGeneric)
+                          .getJson(JsonOptions::none);
+        }
+        catch (std::runtime_error& e)
+        {
+            return RPC::invalid_field_error(jss::tx_blob);
+        }
+    }
+    else if (context.params.isMember(jss::tx_json))
+    {
+        tx_json = context.params[jss::tx_json];
+        if (!tx_json.isObject())
+        {
+            return RPC::object_field_error(jss::tx_json);
+        }
+    }
+    else
+    {
+        // neither `tx_blob` nor `tx_json` included
+        return rpcError(rpcINVALID_PARAMS);
+    }
+
+    // basic sanity checks for transaction shape
+    if (!tx_json.isMember(jss::TransactionType))
+    {
+        return RPC::missing_field_error("tx.TransactionType");
+    }
+    if (!tx_json.isMember(jss::Account))
+    {
+        return RPC::missing_field_error("tx.Account");
+    }
+
+    // autofill fields if they're not included (e.g. `Fee`, `Sequence`)
+    if (auto error = autofillTx(tx_json, context))
+        return *error;
+
+    STParsedJSONObject parsed(std::string(jss::tx_json), tx_json);
+    if (!parsed.object.has_value())
+    {
+        jvResult[jss::error] = parsed.error[jss::error];
+        jvResult[jss::error_code] = parsed.error[jss::error_code];
+        jvResult[jss::error_message] = parsed.error[jss::error_message];
+        return jvResult;
+    }
+
+    std::shared_ptr<STTx const> stpTrans;
+    try
+    {
+        stpTrans = std::make_shared<STTx>(std::move(parsed.object.value()));
+    }
+    catch (std::exception& e)
+    {
+        jvResult[jss::error] = "invalidTransaction";
+        jvResult[jss::error_exception] = e.what();
+
+        return jvResult;
+    }
+
+    std::string reason;
+    auto tpTrans = std::make_shared<Transaction>(stpTrans, reason, context.app);
```

Comment:
`tpTrans` is unclear name, could you please rename it

Reply:
Was due to copy-paste, fixed in 0bbf589f53a13069448d709faa5cea88b6ee547f